File: packages/nitro-wind/cpp/engine/Engine.cpp
#include "Engine.hpp"
#include "Hash.hpp"
#include "Inflate.hpp"
#include "Parser.hpp"
#include "Theme.hpp"

namespace nitrowind::engine {

StyleResult Engine::compute(const std::string& className, const EngineContext& context) {
  const uint64_t key = cacheKey(className, contextBitmask(context), themeHash_);
  if (hasLast_ && key == lastKey_) {
    return lastResult_;
  }
  if (const auto* cached = cache_.get(key)) {
    hasLast_ = true;
    lastKey_ = key;
    lastResult_ = *cached;
    return lastResult_;
  }
  const auto parsed = parseResolved(className, context);
  StyleResult computed;
  computed.style = inflateStyle(parsed.style);
  computed.animation = parsed.animation;
  cache_.set(key, computed);
  hasLast_ = true;
  lastKey_ = key;
  lastResult_ = std::move(computed);
  return lastResult_;
}

std::vector<StyleResult> Engine::computeBatch(const std::vector<std::string>& classNames, const EngineContext& context) {
  std::vector<StyleResult> results;
  results.reserve(classNames.size());
  for (const auto& className : classNames) {
    results.push_back(compute(className, context));
  }
  return results;
}

void Engine::setThemeName(const std::string& name) {
  if (name == themeName_) return;
  themeName_ = name;
  themeHash_ = (name.empty() || name == "default") ? 0 : fnv1a64(name);
  hasLast_ = false;
  lastKey_ = 0;
  lastResult_ = {};
}

void Engine::registerThemeTokens(const std::string& payload) {
  ::nitrowind::engine::registerThemeTokens(payload);
  themeHash_ ^= fnv1a64(payload) | 1;
  clearCache();
}

void Engine::clearCache() {
  cache_.clear();
  hasLast_ = false;
  lastKey_ = 0;
  lastResult_ = {};
}

double Engine::getCacheSize() const {
  return static_cast<double>(cache_.size());
}

} // namespace nitrowind::engine

File: packages/nitro-wind-core/src/engine.ts
import { StyleCache, fastCacheKey } from "./cache";
import { inflateStyle } from "./inflate";
import { parseAnimation, parseClassName } from "./parser";
import { onThemeTokensChanged } from "./theme";
import { type AnimationMeta, DEFAULT_STYLE_CONTEXT, type StyleContext } from "./types";

export interface StyleResult {
	style: Record<string, unknown>;
	animation: AnimationMeta;
}

function isEmptyPartial(context: Partial<StyleContext>): boolean {
	for (const key in context) {
		if (context[key as keyof StyleContext] !== undefined) return false;
	}
	return true;
}

function resolveContext(context: Partial<StyleContext>): StyleContext {
	if (isEmptyPartial(context)) return DEFAULT_STYLE_CONTEXT;
	return { ...DEFAULT_STYLE_CONTEXT, ...context };
}

export class JsStyleEngine {
	private readonly cache = new StyleCache<StyleResult, string>();
	private themeName = "default";
	private lastKey: string | undefined;
	private lastResult: StyleResult | undefined;

	constructor() {
		onThemeTokensChanged(() => {
			this.resetHotCache();
		});
	}

	compute(className: string, context: Partial<StyleContext> = DEFAULT_STYLE_CONTEXT): StyleResult {
		const resolved =
			context === DEFAULT_STYLE_CONTEXT ? DEFAULT_STYLE_CONTEXT : resolveContext(context);
		const key = fastCacheKey(className, resolved);
		if (this.lastKey === key && this.lastResult) {
			return this.lastResult;
		}
		const cached = this.cache.get(key);
		if (cached) {
			this.lastKey = key;
			this.lastResult = cached;
			return cached;
		}
		const result: StyleResult = {
			style: inflateStyle(parseClassName(className, resolved)),
			animation: parseAnimation(className),
		};
		this.cache.set(key, result);
		this.lastKey = key;
		this.lastResult = result;
		return result;
	}

	computeBatch(
		classNames: string[],
		context: Partial<StyleContext> = DEFAULT_STYLE_CONTEXT,
	): StyleResult[] {
		return classNames.map((className) => this.compute(className, context));
	}

	setThemeName(name: string): void {
		if (name === this.themeName) return;
		this.themeName = name;
		this.resetHotCache();
	}

	getThemeName(): string {
		return this.themeName;
	}

	clearCache(): void {
		this.cache.clear();
		this.resetHotCache();
	}

	getCacheSize(): number {
		return this.cache.size;
	}

	private resetHotCache(): void {
		this.cache.clear();
		this.lastKey = undefined;
		this.lastResult = undefined;
	}
}

export const jsStyleEngine = new JsStyleEngine();
